src/core/macro.scala

```diff
@@ -259,12 +281,223 @@ object Macro:
+        val annTerms: List[(String, List[Term])] = 
+          typeRepr.baseClasses
+          .filterNot(isObjectOrScala)
+          .collect {
+            case s if s != typeRepr.typeSymbol =>
+              (fromConstructor(s) ++ fromDeclarations(s)).filter {
+                case (_, anns) => anns.nonEmpty
+              }
+            }
+          .flatten
+        groupTermsByName(annTerms)
+
+      def accumulateTypeAnnotations(typeRepr: TypeRepr): List[Term] = typeRepr match
+        case AnnotatedType(inner, ann) => ann :: accumulateTypeAnnotations(inner)
+        case _                         => Nil
+
+      def getTypeAnnsDict(typeRepr: TypeRepr): List[(String, List[Term])] = 
+        typeRepr
+          .typeSymbol
+          .caseFields
+          .map { field => 
+            val tpeRepr = field.tree match
+              case v: ValDef => v.tpt.tpe
+              case d: DefDef => d.returnTpt.tpe
+            
+            field.name -> accumulateTypeAnnotations(tpeRepr).filter { a =>
+              a.tpe.typeSymbol.maybeOwner.isNoSymbol ||
+              a.tpe.typeSymbol.owner.fullName != "scala.annotation.internal"
+            }
+
+          }
+          .filter(_._2.nonEmpty)
+
+      def getDefaultValue[P: Type](dict: Map[String, Expr[Any]], name: String): Expr[CallByNeed[Option[Any]]] = 
+        dict.get(name) match 
+          case Some(e) => '{ new CallByNeed(() => Some( ($e).asInstanceOf[P] )) }
+          case None => '{ new CallByNeed(() => None) }
+
+      extension (term: Term)
+        def asCallByNeed(tpe: TypeRepr): Term = 
+          Select(
+            '{CallByNeed}.asTerm, 
+            TypeRepr.of[CallByNeed.type].termSymbol.declaredMethod("apply").head
+            )
+            .appliedToType(tpe)
+            .appliedTo(term)
+
+      extension (tpe: TypeRepr)
+        def defaultValues: Map[String, Expr[Any]] = 
+          val sym = tpe.typeSymbol
+          val compModule = Ref(sym.companionModule)
+          val names = 
+            for 
+              a <- sym.caseFields if a.flags.is(Flags.HasDefault)
+            yield a.name
+          val statements = sym.companionClass.tree.asInstanceOf[ClassDef].body
+          val refs: List[Ref] = 
+            for 
+              case defdef @ DefDef(name, _, _, _) <- statements if name.startsWith("$lessinit$greater$default")
+            yield compModule.select(defdef.symbol)
+
+          val refsTerms = refs.map(_.asExpr)
+
+          (names zip refsTerms).toMap 
+
+      extension [B: Type](e: Expr[B])
+        def asCallByNeedExpr: Expr[CallByNeed[B]] = 
+          '{ new CallByNeed[B](() => $e) }
+          
+
+      val aTpe: TypeRepr = TypeRepr.of[A]
+      val aSym: Symbol = aTpe.typeSymbol
+      val aCtor: Symbol = aSym.primaryConstructor
+
+      val defaultValuesDict = aTpe.defaultValues
+      val annsDict = getAnnsDict(aTpe)
+      val inheritedAnnsDict = getInheritedAnnsDict(aTpe)
+      val typeAnnsDict = getTypeAnnsDict(aTpe)
```

Comment:
why not convert to a map once, instead of on every usage? if it's a dict

Reply:
Refactored.